nixos/modules/services/hardware/pid-fan-controller.nix

```
@@ -0,0 +1,163 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.pid-fan-controller;
+  heatSource = {
+    options = {
+      name = lib.mkOption {
+        type = lib.types.uniq lib.types.nonEmptyStr;
+        description = "name of heat source";
+      };
+      wildcard_path = lib.mkOption {
+        type = lib.types.nonEmptyStr;
+        description = ''Wildcard path of heat source hwmon "temp_input", can contain wildcards.'';
+      };
+      PID_params = {
+        set_point = lib.mkOption {
+          type = lib.types.ints.unsigned;
+          description = "set point of controller in °C";
+        };
+        P = lib.mkOption {
+          description = "K_p of PID controller";
+          type = lib.types.float;
+        };
+        I = lib.mkOption {
+          description = "K_i of PID controller";
+          type = lib.types.float;
+        };
+        D = lib.mkOption {
+          description = "K_d of PID controller";
+          type = lib.types.float;
+        };
+      };
+    };
+  };
+
+  fan = {
+    options = {
+      wildcard_path = lib.mkOption {
+        type = lib.types.str;
+        description = ''
+          Wildcard path of the hwmon "pwm" file.
+          If the fans are not to be found in /sys/class/hwmon/hwmon* the corresponding
+          kernel module (like "nct6775") needs to be added to `boot.kernelModules`.
+          See: [hwmon](https://www.kernel.org/doc/html/latest/hwmon/index.html)
+        '';
+      };
+      min_pwm = lib.mkOption {
+        default = 0;
+        type = lib.types.ints.between 0 255;
```

```suggestion
        type = lib.types.ints.u8;
```

Please see https://nixos.org/manual/nixos/unstable/#sec-option-types